client.go

```diff
@@ -0,0 +1,71 @@
+package hlb
+
+import (
+	"context"
+	"net"
+
+	"github.com/docker/buildx/store/storeutil"
+	"github.com/docker/buildx/util/imagetools"
+	dockercommand "github.com/docker/cli/cli/command"
+	"github.com/docker/cli/cli/flags"
+	"github.com/moby/buildkit/client"
+	"github.com/openllb/hlb/codegen"
+	"github.com/openllb/hlb/solver"
+)
+
+// Client returns a BuildKit client specified by addr based on BuildKit's
+// connection helpers.
+//
+// If addr is empty, an attempt is made to connect to docker engine's embedded
+// BuildKit which supports a subset of the exporters and special `moby`
+// exporter.
+func Client(ctx context.Context, addr string) (*client.Client, context.Context, error) {
+	var (
+		dockerCli *dockercommand.DockerCli
+		auth      imagetools.Auth
+		err       error
+	)
+	// Attempt to connect to a healthy docker engine.
+	for _, f := range []func() error{
```

Comment:
Never seen this pattern before. Do you think it's better than the normal `if err := foo(); err != nil { return err }` approach?

Reply:
It evolved into something weird, I fixed it here: https://github.com/openllb/hlb/pull/290/commits/7296571611076ca7b8bba03c5ea3e31b724f1bbe